### PR TITLE
feat: optimize HTML text extraction in TableWidgetV2

### LINKED ISSUE #38275: [Bug]: [HTML Column Type - Table] - Movies db - genres column - if we change type to HTML, all data is lost
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Generate page with data with movies mongo mockdb. For the genres column, change the column type to html. Note that all the data in the table disappears

### Steps To Reproduce

1. Generate page with data using mongo mock db
2. For the genres column, change the column type to html. 
3. Note that all the data in the table disappears
For any other column type, the data is retained .. only if we change it to html, this issue is seen

### Public Sample App

_No response_

### Environment

Release

### Severity

High (Blocker to building or releasing)

### Issue video log

https://jam.dev/c/20df35b3-4ed1-498b-90e0-57f08ca0d737

### Version

Cloud

### PR DESCRIPTION
**Pull request by rahulbarwal**
## Description
- Enhanced the getTextFromHTML function with error handling to prevent failures when processing invalid HTML.
- Introduced logic to conditionally extract HTML text only when necessary, improving performance.
- Updated sorting and filtering mechanisms to utilize extracted HTML text, ensuring accurate comparisons and display.

Fixes #38275
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Table"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12443677014>
> Commit: b98e9a36451c4f7e94a8e5b2a3c848ddfcd9b24a
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12443677014&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Table`
> Spec:
> <hr>Sat, 21 Dec 2024 09:18:53 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling for HTML content processing.
	- Optimized filtering to reduce unnecessary processing of HTML columns.
	- Improved sorting accuracy when handling HTML content.

- **Bug Fixes**
	- Refined logic for extracting and managing HTML text to ensure proper filtering and sorting.

### CODE REVIEW

File: app/client/src/widgets/TableWidgetV2/widget/derived.js [hunk 1]
```diff
@@ -284,13 +284,40 @@ export default {
...
+     * Since getTextFromHTML is an expensive operation, we need to avoid calling it unnecessarily
+     * This optimization ensures that getTextFromHTML is only called when required
+     */
+    const columnsWithHTML = Object.values(props.primaryColumns).filter(
+      (column) => column.columnType === "html",
+    );
+    const htmlColumnAliases = columnsWithHTML.map((column) => column.alias);
+
+    const isFilteringByColumnThatHasHTML = props.filters?.some((filter) =>
+      htmlColumnAliases.includes(filter.column),
```
Comment (vsvamsi1) [inline app/client/src/widgets/TableWidgetV2/widget/derived.js:308]: Can we make the code a little bit more optimum, can htmlColumnAliases be changed to a Set

File: app/client/src/widgets/TableWidgetV2/widget/derived.js [hunk 2]
```diff
@@ -875,12 +921,18 @@ export default {
...
             const originalColValue = isHTMLColumn
-              ? getTextFromHTML(originalRow[props.filters[i].column])
+              ? originalRow[
+                  getKeyForExtractedTextFromHTML(props.filters[i].column)
+                ] ?? getTextFromHTML(originalRow[props.filters[i].column])
               : originalRow[props.filters[i].column];
             const displayedColValue = isHTMLColumn
-              ? getTextFromHTML(displayedRow[props.filters[i].column])
+              ? displayedRow[
+                  getKeyForExtractedTextFromHTML(props.filters[i].column)
```
Comment (vsvamsi1) [inline app/client/src/widgets/TableWidgetV2/widget/derived.js:934]: getKeyForExtractedTextFromHTML(props.filters[i].column) seems to be computed twice , can it be done once and the result be referenced in both places.

File: app/client/src/widgets/TableWidgetV2/widget/derived.js (hunk 1 above)
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/widget/derived.js]: Done.

**Review by jacquesikot — APPROVED** [review]

**Pull request merged**